Add Fossbeamer default config to command line

freyacodes opened the pull request:
Loads a config file into the nix store and provides it to fossbeamer.

The current default isn't final. I don't know what the actual host will be.

We can also ignore the config and put an example file instead.

freyacodes committed:
Add Fossbeamer default config to command line

Co-Authored-By: Florian Klink <flokli@flokli.de>

flokli commented on nix/configuration.nix:
```diff
@@ -62,11 +62,13 @@
   # And configure cage to show fossbeamer on it.
-  services.cage = {
+  services.cage = let
+    defaultConfig = pkgs.writeText "fossbeamer-default-config.json" (builtins.toJSON (import ./default-config.nix));
```
Can we just put the config file as literal JSON in the repo (maybe just in the root)? Going that intermediate step in nixlang feels a bit too much.

flokli commented on nix/configuration.nix:
```diff
@@ -62,11 +62,13 @@
     environment.GDK_BACKEND = "x11";
     user = "kiosk";
-    program = "${pkgs.fossbeamer}/bin/fossbeamer https://example.com";
+    program = "${pkgs.fossbeamer}/bin/fossbeamer --default-config=${defaultConfig} https://example.com";
```
```suggestion
    program = "${pkgs.fossbeamer}/bin/fossbeamer --default-config=${../default-config.json}";
```

flokli commented:
Will this even work without the functionality in the binary?

freyacodes commented:
> Will this even work without the functionality in the binary?

No, @VanuPhantom is doing that separately. That is why is why it created a new branch

flokli reviewed (approved):
Though I feel like this should be cherry-picked into the other branch, or both in the same PR, as it creates breakage as-is.

freyacodes commented:
After talking with @VanuPhantom we decided that I would give it a shot

freyacodes committed:
Support '--default-config=' flag in Rust

flokli commented on src/config.rs:
```diff
@@ -12,12 +13,24 @@ pub(crate) struct Config {
+    pub fn load(default_config_path: Option<PathBuf>) -> Result<Self, Error> {
+        let override_config_path = PathBuf::from("config.json");
+
+        let final_path = if override_config_path.is_file() {
```
Use [`Path::exists`](https://doc.rust-lang.org/std/path/struct.Path.html#method.exists) here, so this can be a symlink too.

freyacodes replied:
[`is_file()`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.is_file) already traverses symlinks

flokli replied:
Ok

flokli commented on src/main.rs:
```diff
@@ -35,3 +38,25 @@ fn main() -> wry::Result<()> {
+    let mut default_config_path: Option<PathBuf> = None;
+    let mut address: Option<String> = None;
+    for string in env::args() {
+        if string.starts_with("http") {
```
Hurgh, what about [clap-derive](https://docs.rs/clap/latest/clap/_derive/index.html#overview)?

freyacodes replied:
Never heard of it. I'll look into it

VanuPhantom replied:
I second @flokli, let's see what this'll look like with derivations.

freyacodes committed:
Use Clap to handle command line arguments

Now with a hopefully happy crate2nix

...and fewer complaints from treefmt over generated code

VanuPhantom commented on src/config.rs:
```diff
@@ -12,12 +13,24 @@ pub(crate) struct Config {
 impl Config {
-    pub fn load() -> Result<Self, Error> {
-        let Ok(file) = File::open("config.json") else {
+    pub fn load(default_config_path: Option<PathBuf>) -> Result<Self, Error> {
+        let override_config_path = PathBuf::from("config.json");
+
+        let final_path = if override_config_path.is_file() {
+            override_config_path
+        } else if let Some(path) = default_config_path {
+            if !path.is_file() {
+                return Err(Error::NoConfigFileFound);
+            }
+            path
+        } else {
+            return Err(Error::NoConfigFileFound);
+        };
+
+        let Ok(file) = File::open(final_path) else {
             return Err(Error::FileIoError);
         };
```
```suggestion
    pub fn load(default_config_path: Option<PathBuf>) -> Result<Self, Error> {
        let config_path = {
            let override_config_path = PathBuf::from("config.json");

            if override_config_path.is_file() {
                Some(override_config_path)
            } else {
                None
            }
        }
        .or(default_config_path)
        .ok_or(Error::NoConfigFileFound)?;

        let file = File::open(config_path).or(Err(Error::FileIoError))?;
```

VanuPhantom reviewed (changes requested):
I see some readability improvements that can be made in `config.rs`. Besides this, I'll wait to see what derivations will do to the code volume in `main.rs`.

freyacodes committed:
Refactor `Config::load()`

Co-authored-by: Tess Ellenoir Duursma <tess@vanu.dev>

VanuPhantom reviewed (approved):
I just reviewed this again, it looks good. I like the elegance of `clap-derive`.
